Update KIC docs to reflect webhook configuration changes (#111)
nginxinc/nginx-service-mesh: docs/content/tutorials/kic/deploy-with-kic.md

> 

Reviewer: I think it's worth mentioning somewhere what our mesh actually does to integrate ingress controller. It's important for users to know:
- adding a label to NIC that hasn't been created yet will result in NSM updating it when it's created
- adding a label to NIC that already exists requires a rollout restart for NSM to pick it up and update it

It's also probably worth mentioning which fields we add to the Pod, so a user knows what we're exactly changing.

Author: I added an initial section explaining what our webhook modifies. I think it gives a pretty detailed explanation, but let me know any changes you would like to see.

{{< important >}}
There are two versions of NGINX Ingress Controller for Kubernetes: NGINX Open Source and NGINX Plus.
To deploy NGINX Ingress Controller with NGINX Service Mesh, you must use either:

> - Open Source NGINX Ingress Controller version 3.0+ or greater

Reviewer 2: Nit: "3.0+" and "or greater" is redundant.

When using NGINX Plus Ingress Controller with mTLS enabled, it is best practice to keep the `ttl` at 1 hour or greater.

## Install NGINX Ingress Controller with mTLS enabled

To configure NGINX Plus Ingress Controller to communicate with mesh workloads over mTLS you need to make a modification to the Ingress Controller's Pod spec. This section describes each modification that is required, but if you'd like to jump to installation, go to the [Install with Manifests](#install-with-manifests) or [Install with Helm](#install-with-helm) sections. 

1. Add NGINX Service Mesh label

>    One or both of the following labels must be added to the Ingress Controller's Pod spec:

Reviewer: ```suggestion
   One or both of the following labels must be added to the Ingress Controller's Pod spec, depending on your use case:
```

    ```yaml
    labels:
      nsm.nginx.com/enable-ingress: "true"
      ...
    ```

    ```yaml
    labels:
      nsm.nginx.com/enable-egress: "true"
      ...
    ```

>    This label prevents NGINX Service Mesh from automatically injecting the sidecar into the Ingress Controller Pod.

Reviewer: ```suggestion
   These labels tell NGINX Service Mesh to mutate the Ingress Controller Pod with the proper configuration in order to properly integrate with the mesh.
```

### Install with Manifests

Before installing NGINX Plus Ingress Controller, you must install NGINX Service Mesh with an [mTLS mode]({{< ref "/guides/secure-traffic-mtls.md" >}}) of `permissive`, or `strict`.

> NGINX Plus Ingress Controller will try to fetch certs from the SPIRE agent on startup. If it cannot reach the SPIRE agent, startup will fail, and NGINX Plus Ingress Controller will go into CrashLoopBackoff state. The state will resolve once NGINX Ingress Controller connects to the SPIRE agent.

Reviewer: ```suggestion
NGINX Ingress Controller will try to fetch certs from the SPIRE agent on startup. If it cannot reach the SPIRE agent, startup will fail, and NGINX Ingress Controller will go into CrashLoopBackoff state. The state will resolve once NGINX Ingress Controller connects to the SPIRE agent.
```

May want to do an audit of all the instances of NGINX Plus Ingress Controller, there may be more.

For instructions on how to install NGINX Service Mesh, see the [Installation]({{< ref "/get-started/install.md" >}}) guide.

{{< note >}}
Before continuing, check the NGINX Ingress Controller [supported versions](#supported-versions) section and make sure you are working off the correct release tag for all NGINX Ingress Controller instructions.
{{< /note >}}

> 1. Build or Pull the NGINX OSS/Plus Ingress Controller image:

Reviewer: Since these instructions are very tailored for the Plus version, it may just make more sense to have a different section for OSS. Does a user even need to build OSS? Plus is special since a user has to build it themselves with their cert/key for N+

Author: I broke every occurrence of this into separate sections.

        - Add the `nginx-ingress-permissions` to the ServiceAccount of the NGINX Ingress Controller.

          ```bash
          oc adm policy add-scc-to-user nginx-ingress-permissions -z nginx-ingress -n nginx-ingress
          ```

#### NGINX Plus Ingress Controller

1. Build or Pull the NGINX Plus Ingress Controller image:
    - [Create and push your NGINX Plus Docker image](https://docs.nginx.com/nginx-ingress-controller/installation/building-ingress-controller-image/).
    - For NGINX Plus Ingress releases >= `1.12.0` you can also [pull the NGINX Plus Docker image](https://docs.nginx.com/nginx-ingress-controller/installation/pulling-ingress-controller-image/).
1. Set up Kubernetes Resources for [NGINX Plus Ingress Controller](https://docs.nginx.com/nginx-ingress-controller/installation/installation-with-manifests/) using Kubernetes manifests:
    - [Configure role-based access control (RBAC)](https://docs.nginx.com/nginx-ingress-controller/installation/installation-with-manifests/#1-configure-rbac)
    - [Create Common Resources](https://docs.nginx.com/nginx-ingress-controller/installation/installation-with-manifests/#2-create-common-resources)

> 1. Create the NGINX Ingress Controller as a **Deployment** or **DaemonSet** in Kubernetes using one of the following example manifests:

Reviewer: This could stay as Plus

NGINX Ingress Controller will try to fetch certs from the SPIRE agent on startup. If it cannot reach the SPIRE agent, startup will fail, and NGINX Ingress Controller will go into CrashLoopBackoff state. The state will resolve once NGINX Ingress Controller connects to the SPIRE agent.
For instructions on how to install NGINX Service Mesh, see the [Installation]({{< ref "/get-started/install.md" >}}) guide.
   
{{< note >}} NGINX Plus Ingress Controller v2.2+ or NGINX OSS Ingress Controller v3.0+ is required to deploy via Helm and integrate with NGINX Service Mesh. {{< /note >}}

Follow the [instructions](https://docs.nginx.com/nginx-ingress-controller/installation/installation-with-helm/) to install the NGINX Ingress Controller with Helm.
Set the `nginxServiceMesh.enable` parameter to `true`.
{{< note >}} This will configure NGINX Ingress Controller to route ingress traffic to NGINX Service Mesh workloads. If you would like to enable egress traffic, refer to the [Enable Egress](#enable-with-helm) section of this guide. {{< /note >}}

> The [`values-nsm.yaml`](https://github.com/nginxinc/kubernetes-ingress/blob/master/deployments/helm-chart/values-nsm.yaml) file contains all the configuration parameters that are relevant for integration with NGINX Service Mesh. You can use this file if you are installing NGINX Ingress Controller via chart sources.

Reviewer: May not really need this anymore, or it should just include the necessary labels for integration.

Reviewer: Per offline talk, this will stay for now.

    - For *Deployment*:

      ```yaml
      nsm.nginx.com/deployment: <name of NGINX Plus Ingress Controller Deployment>
      nsm.nginx.com/deployment: <name of NGINX Ingress Controller Deployment>
      ```

    - For *DaemonSet*:

      ```yaml
      nsm.nginx.com/daemonset: <name of NGINX Plus Ingress Controller DaemonSet>
      nsm.nginx.com/daemonset: <name of NGINX Ingress Controller DaemonSet>
      ```

    This allows metrics scraped from NGINX Plus Ingress Controller Pods to be associated with the resource that created the Pods.
    This allows metrics scraped from NGINX Ingress Controller Pods to be associated with the resource that created the Pods.

### View the metrics in Prometheus

The NGINX Service Mesh uses the Pod's container name setting to identify the NGINX Plus Ingress Controller metrics that should be consumed by the Prometheus server.
The NGINX Service Mesh uses the Pod's container name setting to identify the NGINX Ingress Controller metrics that should be consumed by the Prometheus server.
The Prometheus job targets all Pods that have the container name `nginx-plus-ingress`.

> Add the `nginx-plus-ingress` scrape config to your Prometheus configuration and consult

Reviewer: We probably want two sections for each of the two types of scrape configs. I see you added an OSS scrape config.

      nsm.nginx.com/daemonset: <name of NGINX Ingress Controller DaemonSet>
      ```

   This allows metrics scraped from NGINX Ingress Controller Pods to be associated with the resource that created the Pods.

> #### Enabling NGINX Plus Ingress Controller Metrics

Reviewer: I think we can just remove this entire section. In the section above, simply add a note following the first step that says something like "If using N+ controller, add this additional flag (enable-latency-metrics). Everything else between these sections is the same.

Here is a view of the provided "NGINX Mesh Top" dashboard:

{{< img src="/img/grafana.png" >}}

> ## Integrating NGINX Ingress Controller with NSM

Reviewer: NSM shouldn't be used in our public docs. Be sure to use the full NGINX Service Mesh phrasing (and be sure to update the link to this section.

> ## Integrating NGINX Ingress Controller with NSM

Reviewer 2: Typically we use "NGINX Service Mesh" instead of "NSM" in public docs.
